File: Data_Structure/Linked_List/Circular_Singly_Linked_List.java
/*
Linked List: Circular Singly Linked List
*/

package Data_Structure.Linked_List;

public class Circular_Singly_Linked_List {

    private static class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    public static void main(String[] args) {
        Node  head = new Node(10);
        Node node1 = new Node(20);
        Node node2 = new Node(30);
        Node tail = new Node(40);

        head.next = node1;
        node1.next = node2;
        node2.next = tail;
        tail.next =  head; // circular link

        Node currNode =  head;

        System.out.print(currNode.data  + " -> ");
        currNode = currNode.next;
        while (currNode != head) {
            System.out.print(currNode.data  + " -> ");
            currNode = currNode.next;
        }
        System.out.println("...");
    }
}

File: Data_Structure/Linked_List/Circular_Singly_Linked_List.cpp
// Circular Singly Linked List

#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
};

void insert(Node*& head, int value) {
    Node* newNode = new Node{value, nullptr};
    if(!head) {
        head = newNode;
        newNode -> next = head;
    } else {
        Node* temp = head;
        while(temp -> next != head) {
            temp = temp -> next;
        }
        temp -> next = newNode;
        newNode -> next = head;
    }
}

void deletelist(Node*& head) {
    if(!head) return;
    Node* temp = head;
    do {
        Node* next = temp -> next;
        delete temp;
        temp = next;
    } while(temp != head);
    head = nullptr;
}

void display(Node* head) {
    if(!head) return;
    Node* temp = head;
    do {
        cout << temp -> data << " -> ";
        temp = temp -> next;
    } while(temp != head);
    cout << "back to head" << endl;
}

int main() {

    Node* head = nullptr;
    insert(head, 1);
    insert(head, 2);
    insert(head, 3);

    display(head);
    deletelist(head);
    if(!head) {
        cout << "list delete successfully" << endl;
    }

    return 0;
}